#!/usr/bin/env Rscript

# Author: Robert Wang (Xing Lab)
# Date: 2024.09.21
# Version: 1.0.0

# A script designed to compute the on-target rate for long-read RNA sequencing data given a BED file with target genes. 
# This script also identifies the top N most abundant genes from read alignments.

# =====================================================================================================================
#                                                      LIBRARIES 
# =====================================================================================================================

# Load required libraries
suppressMessages(library(argparse))
suppressMessages(library(dplyr))
suppressMessages(library(ggplot2))
suppressMessages(library(tidyr))

# =====================================================================================================================
#                                                   HELPER FUNCTIONS
# =====================================================================================================================

PullFeature <- function(infoString, featureName) {
    # Function designed to pull out the value for featureName in infoString
    present <- unlist(lapply(strsplit(infoString, "; "), function(x) lapply(strsplit(x, " "), "[[", 1) == featureName))
    return(ifelse(sum(present) == 1, unlist(lapply(strsplit(unlist(strsplit(infoString, "; "))[present], " "), "[[", 2)), NA))
}

# =====================================================================================================================
#                                                    MAIN FUNCTIONS
# =====================================================================================================================

message <- "Computes on-target rate and identifies the top N most abundant genes from long-read RNA sequencing alignments"
parser <- ArgumentParser(description = message)

# Add arguments to parser
parser$add_argument("--gtffile", metavar = "/path/to/stringtie/GTF/file",
    help = "path to GTF file generated by stringtie on input read alignments")
parser$add_argument("--gene_abundance", metavar = "/path/to/stringtie/gene/abundance/file",
    help = "path to gene abundance file generated by stringtie on input read alignments")
parser$add_argument("--targets", metavar = "/path/to/target/gene/BED/file",
    help = "path to BED file with genomic coordinates of target genes")
parser$add_argument("--count", default = 2000, type = "integer",
    help = "number of most abundant genes to report")
parser$add_argument("--stats", metavar = "/path/to/mapping/stats/file",
    help = "path to output file with read mapping statistics")
parser$add_argument("--gene_output", metavar = "/path/to/gene/output/file",
    help = "path to output file with top N most abundant genes")
parser$add_argument("--plot_pdf", metavar = "/path/to/gene/abundance/plot/pdf",
    help = "path to output PDF plot for top N most abundant genes")

# Parse command-line arguments
args <- parser$parse_args()

# Determine which STRG IDs are matched to reference gene IDs
mapping <- read.table(args$gtffile, sep = "\t", header = FALSE, comment = "#") %>% filter(V3 == "transcript") %>%
    mutate(Gene_ID = sapply(V9, function(x) PullFeature(x, "gene_id")), 
    Reference_Gene_ID = sapply(V9, function(x) PullFeature(x, "ref_gene_id"))) %>%
    select(Gene_ID, Reference_Gene_ID) %>% drop_na()

# Annotate genes in gene_abundance based on whether they are target genes or not
targets <- suppressWarnings(read.table(args$targets, header = FALSE) %>% separate(V4, c("V4", NA), sep = "\\.")) %>% pull(V4)
outDF <- read.table(args$gene_abundance, sep = "\t", header = TRUE) %>% filter(!(Gene.ID %in% mapping$Gene_ID)) %>% 
    mutate(Group = case_when(substr(Gene.ID, 1, 15) %in% targets ~ "Target", TRUE ~ "Non-target")) %>%
    select(Gene.ID, Group, TPM) %>% mutate(Rank = rank(-TPM, ties.method = "first"), 
    Group = factor(Group, levels = c("Target", "Non-target")))

# Compute on-target rate and append to stats
on.target.rate <- paste(formatC(round(sum(outDF[outDF$Group == "Target", ]$TPM)/1e4, digits = 2), format = "f", flag = "0", 
    digits = 2), "%", sep = "")
write(paste(c(on.target.rate, "N/A", "on-target rate %"), collapse = "\t"), file = args$stats, append = TRUE)

# Filter outDF for the top N most abundant genes and save to gene_output
outDF <- outDF[outDF$Rank <= args$count, ]
write.table(select(outDF, -Rank) %>% arrange(-TPM), file = args$gene_output, quote = FALSE, sep = "\t", row.names = FALSE)

# Plot gene rank versus gene abundance for genes in outDF
p <- ggplot(outDF, aes(x = Rank, y = log2(TPM + 1))) + theme_classic() + geom_bar(aes(fill = Group), linewidth = 0, stat = "identity") + 
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank(), 
    axis.text = element_text(color = "black", size = 5), axis.title = element_text(color = "black", size = 6), 
    axis.ticks = element_line(color = "black", linewidth = 0.25), legend.title = element_blank(), legend.text = element_text(color = "black", size = 5), 
    legend.key.size = unit(0.15, 'cm')) + xlab("Genes ranked by abundance") + ylab("Gene abundance\nlog2(TPM+1)") +
    annotate("text", x = 0.25 * args$count, y = 0.75 * max(log2(outDF$TPM) + 1), label = paste("On-target rate:", on.target.rate), hjust = 0, size = 5 * 5/14) +
    scale_fill_manual(values = setNames(c("#0571B0", "gray80"), c("Target", "Non-target")))

# Save p to plot_pdf
ggsave(args$plot_pdf, plot = p, width = 2.5, height = 1.25)
